### PR TITLE
chore: Dashboard cypress refactor

### PR DESCRIPTION
**Pull request by adam-stasiak**
### SUMMARY

I closed old PR (I thought I can reopen when ready but I forced push changes with data-test attributes from https://github.com/apache/incubator-superset/pull/11254 -then it was blocked by github) and resolved review suggestions from https://github.com/apache/incubator-superset/pull/11049.
I moved dashboard tests to use data-test attributes.
### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset-frontend/cypress-base/cypress/integration/dashboard/edit_mode.test.js [hunk 1]
```diff
@@ -23,47 +23,54 @@ describe('Dashboard edit mode', () => {
...
     cy.login();
     cy.visit(WORLD_HEALTH_DASHBOARD);
-    cy.get('.dashboard-header [data-test=edit-alt]').click();
+    cy.get('[data-test="dashboard-header"]')
+      .find('[data-test=edit-alt]')
+      .click();
   });
 
-  xit('remove, and add chart flow', () => {
+  it('remove, and add chart flow', () => {
```
Comment (rusackas) [inline superset-frontend/cypress-base/cypress/integration/dashboard/edit_mode.test.js:31]: Happy to see this test making a comeback! I hope it works more reliably now! :)

File: superset-frontend/cypress-base/cypress/integration/dashboard/edit_mode.test.js [hunk 2]
```diff
@@ -23,47 +23,54 @@ describe('Dashboard edit mode', () => {
...
     // wait for box plot to appear
-    cy.get('.grid-container .box_plot');
-
-    cy.get('.fa.fa-trash')
+    cy.get('[data-test="grid-container"]').find('.box_plot', {
+      timeout: 10000,
+    });
+    cy.get('[data-test="dashboard-component-chart-holder"]')
+      .find('[data-test="dashboard-delete-component-button"]')
```
Comment (rusackas) [inline superset-frontend/cypress-base/cypress/integration/dashboard/edit_mode.test.js:37]: Sorry if I've asked this before, but I'm curious why we need to `get` AND `find` here. If `dashboard-delete-component-button` is specific enough, it seems like we could just `get` that directly.

File: superset-frontend/cypress-base/cypress/integration/dashboard/edit_mode.test.js [hunk 3]
```diff
@@ -23,47 +23,54 @@ describe('Dashboard edit mode', () => {
...
+      .find('.box_plot')
+      .should('be.visible');
 
     // should show Save changes button
-    cy.get('.dashboard-header .button-container').contains('Save');
+    cy.get('[data-test="header-save-button"]').should('be.visible');
 
     // undo 2 steps
-    cy.get('.dashboard-header .undo-action').click().click();
+    cy.get('[data-test="undo-action"]').click();
```
Comment (rusackas) [inline superset-frontend/cypress-base/cypress/integration/dashboard/edit_mode.test.js:71]: Not sure why the second `cilck()` was removed here, but the comment above should be updated if this is intended.

More importantly, it's odd that there's no assertion here, like: 
• making sure the box_plot chart is gone again, after clicking the "undo" button.
• making sure the 'save' button is disabled after undoing the changes.

File: superset-frontend/cypress-base/cypress/integration/dashboard/edit_mode.test.js [hunk 4]
```diff
@@ -23,47 +23,54 @@ describe('Dashboard edit mode', () => {
...
 
     // no changes, can switch to view mode
-    cy.get('.dashboard-header .button-container')
-      .contains('Discard Changes')
-      .click();
+    cy.get('[data-test="discard-changes-button"]').click();
```
Comment (rusackas) [inline superset-frontend/cypress-base/cypress/integration/dashboard/edit_mode.test.js:74]: Seems a little odd there's no assertion here either... maybe we should make sure we're not in edit mode anymore?

File: superset-frontend/cypress-base/cypress/integration/dashboard/tabs.test.js [hunk 5]
```diff
@@ -162,7 +176,13 @@ describe('Dashboard tabs', () => {
     });
 
     // click row level tab, send 1 more query
-    cy.get('.tab-content ul.nav.nav-tabs li').last().click();
+    cy.get('[data-test="dashboard-component-tabs"]')
+      .last()
+      .find('[data-test="nav-list"]')
+      .children()
+      .as('row-level-tabs');
```
Comment (rusackas) [inline superset-frontend/cypress-base/cypress/integration/dashboard/tabs.test.js:183]: Appreciate all these `as(...)` changes! Much easier to read, in my book.

**Review by rusackas — COMMENTED** [review]
Couple of questions on the thread, but this is looking great thus far! Obviously there are some CI issues that will need to be investigated as well, so I'll keep an eye on any relevant changes.

File: superset-frontend/cypress-base/cypress/integration/dashboard/edit_mode.test.js (hunk 2 above)
Comment (adam-stasiak) [reply superset-frontend/cypress-base/cypress/integration/dashboard/edit_mode.test.js:37]: To reflect structure of component - I just want to show in test that this delete button belongs to chart holder - it is nothing to make tests more stable - just more strict.

File: superset-frontend/cypress-base/cypress/integration/dashboard/edit_mode.test.js (hunk 4 above)
Comment (adam-stasiak) [reply superset-frontend/cypress-base/cypress/integration/dashboard/edit_mode.test.js]: added assertion for this

File: superset-frontend/cypress-base/cypress/integration/dashboard/tabs.test.js (hunk 5 above)
Comment (adam-stasiak) [reply superset-frontend/cypress-base/cypress/integration/dashboard/tabs.test.js:183]: Yes it is cool!

File: superset-frontend/cypress-base/cypress/integration/dashboard/edit_mode.test.js (hunk 3 above)
Comment (adam-stasiak) [reply superset-frontend/cypress-base/cypress/integration/dashboard/edit_mode.test.js:73]: added assertions and returned click back

File: superset-frontend/cypress-base/cypress/integration/dashboard/edit_mode.test.js (hunk 1 above)
Comment (adam-stasiak) [reply superset-frontend/cypress-base/cypress/integration/dashboard/edit_mode.test.js:31]: I added extra step with filtering first and removed .contains in step back - I had problems with this contains that cypress cannot deal with drag after this.

File: superset-frontend/cypress-base/cypress/integration/dashboard/save.test.js [hunk 6]
```diff
@@ -50,21 +52,32 @@ describe('Dashboard save action', () => {
 
   it('should save/overwrite dashboard', () => {
```
Comment (rusackas) [inline superset-frontend/cypress-base/cypress/integration/dashboard/save.test.js:53]: This test seems to be failing on CI:

Comment (adam-stasiak) [reply superset-frontend/cypress-base/cypress/integration/dashboard/save.test.js:53]: Fixed :)

**Review by rusackas — APPROVED** [review]

**Pull request merged**